reactioncommerce/reaction
# chore: use yalc for package link and unlink

Resolves #6248
Impact: **minor**
Type: **feature|chore**

MERGE FIRST (node-dev image updates) = https://github.com/reactioncommerce/docker-base/pull/31

## Issue

`package-unlink` seems to unlink more packages than it should. This, plus issues with `package-link` itself linking packages, but not properly resolving dependencies which leads to code & tests passing locally, then failing on CI.

## Solution

Use [yalc](https://www.npmjs.com/package/yalc) instead of `npm link`.

Command breakdown

### Linking

Link all packages found in `yalc-packages` where `path/to/package=true`
Unlink all packages found in `yalc-packages` where `path/to/package=false`

```bash
./bin/package-link
```

Link a new package & update an existing package when you've made changes
```bash
./bin/package-link @reactioncommerce/api-plugin-accounts
```

### Unlinking
Unlink a package, run npm install, and re-link previously linked packages
```bash
./bin/package-unlink @reactioncommerce/api-plugin-accounts
```

### Updating
Update linked packages based on the entries from the `yalc.lock` file. If you stopped and started the API container, this will re-link everything.
```bash
./bin/package-update
```

## Breaking changes

None, but...

You should destroy your containers and start fresh. Since this is a dev tool, it won't break dependency management otherwise.

## Testing
1. Checkout the reaction-development-platform
1. Follow setup instructions namely `make init` and `make clone-api-plugin`
1. run the API
1. `cd reaction`
1. `./bin/package-link @reactioncommerce/api-core` or any api plugin in the `api-plugins` directory
1. See the package is installed and running
1. Make the ping query in a GraphQL playgorund (http://localhost:3000/graphql) to see `pong`
1. Make a change or something. I edited `src/graphql/resolvers/index.js` and changed ping to` ping: () => "pong TEST"`
1. run `./bin/package-link @reactioncommerce/api-core` again to update
1. run the `ping` query and see `pong TEST`
1. unlink with `./bin/package-unlink @reactioncommerce/api-core`
1. run the `ping` query again, and you should only see `pong`

---

1. run `./bin/package-link` without any arguments
1. see all packages are linked
1. edit `yalc-packages`
1. set some packages to `false`
1. run `./bin/package-link` again
1. verify that those packages are removed
